mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java

```diff
@@ -0,0 +1,369 @@
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .and(MESSAGE_UID.in(uidsToFetch.stream().map(MessageUid::asLong).toArray(Long[]::new)))
+                .orderBy(DEFAULT_SORT_ORDER_BY)))
+            .map(RECORD_TO_COMPOSED_MESSAGE_ID_WITH_META_DATA_FUNCTION);
+
+        if (messageUids.size() <= IN_CLAUSE_MAX_SIZE) {
+            return queryPublisherFunction.apply(messageUids);
+        } else {
+            return Flux.fromIterable(Iterables.partition(messageUids, IN_CLAUSE_MAX_SIZE))
+                .flatMap(queryPublisherFunction);
+        }
+    }
+
+    public Flux<ComposedMessageIdWithMetaData> findAllRecentMessageMetadata(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.select()
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .and(IS_RECENT.eq(true))
+                .orderBy(DEFAULT_SORT_ORDER_BY)))
+            .map(RECORD_TO_COMPOSED_MESSAGE_ID_WITH_META_DATA_FUNCTION);
+    }
+
+    public Mono<Void> updateFlag(PostgresMailboxId mailboxId, MessageUid uid, UpdatedFlags updatedFlags) {
+        return postgresExecutor.executeVoid(dslContext ->
+            Mono.from(buildUpdateFlagStatement(dslContext, updatedFlags, mailboxId, uid)));
+    }
+
+    public Mono<Flags> listDistinctUserFlags(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(dslContext.selectDistinct(UNNEST_FIELD.apply(USER_FLAGS))
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))))
+            .map(record -> record.get(0, String.class))
+            .collectList()
+            .map(flagList -> {
+                Flags flags = new Flags();
+                flagList.forEach(flags::add);
+                return flags;
+            });
+    }
+
+    private UpdateConditionStep<Record> buildUpdateFlagStatement(DSLContext dslContext, UpdatedFlags updatedFlags,
+                                                                 PostgresMailboxId mailboxId, MessageUid uid) {
+        AtomicReference<UpdateSetStep<Record>> updateStatement = new AtomicReference<>(dslContext.update(TABLE_NAME));
+
+        BOOLEAN_FLAGS_MAPPING.forEach((flagColumn, flagMapped) -> {
+            if (updatedFlags.isChanged(flagMapped)) {
+                updateStatement.getAndUpdate(currentStatement -> {
+                    if (flagMapped.equals(Flag.RECENT)) {
+                        return currentStatement.set(flagColumn, updatedFlags.getNewFlags().contains(Flag.RECENT));
+                    }
+                    return currentStatement.set(flagColumn, updatedFlags.isModifiedToSet(flagMapped));
+                });
+            }
+        });
+
+        return updateStatement.get()
+            .set(USER_FLAGS, updatedFlags.getNewFlags().getUserFlags())
```

Comment:
Please we need flag updates to commute.
Which mean we need a ADD/REMOVE semantic, and not a SET semantic.

Let's SAY IF we only add elements:

```
user_flag = user_flag || ARRAY['nouvel_element1', 'nouvel_element2']
```

Let's SAY IF we only remove elements:

```
user_flag = array_remove(user_flag, ARRAY['nouvel_element1', 'nouvel_element2'])
```

If we both add and remove elements it is OK to do 2 requests...

Reply:
I tried it, but `array_remove` doesn't receive the Array input, 
It just supports "single element"
But yes with the array_concat (for append)

Reply:
what about this in this ticket? https://github.com/linagora/james-project/issues/4958#issuecomment-1829749045